manifest.go

```diff
@@ -0,0 +1,655 @@
+	// will be skipped.
+	FetchEntries(fs iceio.IO, discardDeleted bool) ([]ManifestEntry, error)
+}
+
+// ReadManifestList reads in an avro manifest list file and returns a slice
+// of manifest files or an error if one is encountered.
+func ReadManifestList(in io.Reader) ([]ManifestFile, error) {
+	dec, err := ocf.NewDecoder(in)
+	if err != nil {
+		return nil, err
+	}
+
+	out := make([]ManifestFile, 0)
+
+	for dec.HasNext() {
+		var file ManifestFile
+		if string(dec.Metadata()["format-version"]) == "2" {
+			file = &manifestFileV2{}
+		} else {
+			file = &manifestFileV1{}
+		}
+
+		if err := dec.Decode(file); err != nil {
+			return nil, err
+		}
+		out = append(out, file)
+	}
+
+	return out, dec.Error()
+}
+
+// ManifestEntryStatus defines constants for the entry status of
+// existing, added or deleted.
+type ManifestEntryStatus int8
+
+const (
+	EntryStatusEXISTING ManifestEntryStatus = 0
+	EntryStatusADDED    ManifestEntryStatus = 1
+	EntryStatusDELETED  ManifestEntryStatus = 2
+)
+
+// ManifestEntryContent defines constants for the type of file contents
+// in the file entries. Data, Position based deletes and equality based
+// deletes.
+type ManifestEntryContent int8
+
+const (
+	EntryContentData       ManifestEntryContent = 0
+	EntryContentPosDeletes ManifestEntryContent = 1
+	EntryContentEqDeletes  ManifestEntryContent = 2
+)
+
+// FileFormat defines constants for the format of data files.
+type FileFormat string
+
+const (
+	AvroFile    FileFormat = "AVRO"
+	OrcFile     FileFormat = "ORC"
+	ParquetFile FileFormat = "PARQUET"
+)
+
+type colMap[K, V any] struct {
+	Key   K `avro:"key"`
+	Value V `avro:"value"`
+}
+
+func avroColMapToMap[K comparable, V any](c *[]colMap[K, V]) map[K]V {
+	if c == nil {
+		return nil
+	}
+
+	out := make(map[K]V)
+	for _, data := range *c {
+		out[data.Key] = data.Value
+	}
+	return out
+}
+
+type dataFile struct {
+	Content          ManifestEntryContent   `avro:"content"`
+	Path             string                 `avro:"file_path"`
+	Format           FileFormat             `avro:"file_format"`
+	PartitionData    map[string]any         `avro:"partition"`
+	RecordCount      int64                  `avro:"record_count"`
+	FileSize         int64                  `avro:"file_size_in_bytes"`
+	BlockSizeInBytes int64                  `avro:"block_size_in_bytes"`
+	ColSizes         *[]colMap[int, int64]  `avro:"column_sizes"`
+	ValCounts        *[]colMap[int, int64]  `avro:"value_counts"`
+	NullCounts       *[]colMap[int, int64]  `avro:"null_value_counts"`
+	NaNCounts        *[]colMap[int, int64]  `avro:"nan_value_counts"`
+	DistinctCounts   *[]colMap[int, int64]  `avro:"distinct_counts"`
+	LowerBounds      *[]colMap[int, []byte] `avro:"lower_bounds"`
+	UpperBounds      *[]colMap[int, []byte] `avro:"upper_bounds"`
+	KeyMetadata      *[]byte                `avro:"key_metadata"`
+	SplitOffsets     *[]int64               `avro:"split_offsets"`
+	EqualityIDs      *[]int                 `avro:"equality_ids"`
```

Comment:
this only exists for V2 but not for V1

Reply:
I know, since it is only optional in v2 i'm able to get away without needing separate data structures for V1 and V2 data files since the avro schema will ensure it would be ignored in a v1 file and just left nil. If preferred I can add a comment here to that effect. I just wanted to avoid the duplication since I don't need the separate v1/v2 data file structs.

Reply:
yeah I just wanted to make sure we don't accidentally accept this field when we really shouldn't